### PR TITLE
Deserialize raw byte arrays (packets) into workable datatypes

### PR DESCRIPTION
**Pull request by jdonszelmann**
Converts byte arrays into rust datatypes in a way which is very similar to the way py-ipv8 does it.

### CODE REVIEW

File: examples/connecting.rs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (dsluijk) [inline examples/connecting.rs]: Please remove these unnecessary empty lines

File: src/networking/payload/bits.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+
```
Comment (dsluijk) [inline src/networking/payload/bits.rs]: Please remove these unnecessary empty line

File: src/networking/payload/bits.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#[derive(PartialEq, Debug)]
+pub struct Bits{
```
Comment (dsluijk) [inline src/networking/payload/bits.rs]: What does this exactly do?

File: src/networking/payload/bits.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  bit3: bool,
+  bit4: bool,
+  bit5: bool,
+  bit6: bool,
+  bit7: bool,
+}
+
+
+impl Bits{
+  pub fn from(num:u8) -> Self{
```
Comment (dsluijk) [inline src/networking/payload/bits.rs]: What does this do?

File: src/networking/payload/introductionrequestpayload.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+impl Ipv8Payload for IntroductionRequestPayload {
+  fn pack(&self) -> Packet{
+    return Packet::from(&[1,2])
+  }
+
+  fn unpack(packet: Packet) -> Self{
+    // packet format copied from:
+    // https://github.com/Tribler/py-ipv8/blob/57c1aa73eee8a3b7ee6ad48482fc2e0d5849415e/ipv8/messaging/payload.py#L39
+    packet.unpack(vec![
+      vec![CHAR, CHAR, CHAR, CHAR, U16],
```
Comment (dsluijk) [inline src/networking/payload/introductionrequestpayload.rs:50]: I think it's better to use lowercase for types

File: src/networking/payload/introductionrequestpayload.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      vec![CHAR, CHAR, CHAR, CHAR, U16],
+      vec![BITS], // the connection type and also stores advice in the last bit
+      vec![U16],
+      vec![STRING],
+    ]);
+
+    IntroductionRequestPayload {
+        destination_address : IPAddress {
+          address: String::from("1.2.3.4"),
```
Comment (dsluijk) [inline src/networking/payload/introductionrequestpayload.rs]: Are these supposed to be set like this?

File: src/networking/payload/introductionrequestpayload.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+          port: 6421,
+          version: IPVersion::IPV4,
+        },
+        advice:true,
+        connection_type:ConnectionType::PUBLIC,
+        identifier:1,
+        extra_bytes: String::from("42"),
+    }
+  }
+
```
Comment (dsluijk) [inline src/networking/payload/introductionrequestpayload.rs]: Please remove these unnecessary empty lines

File: src/networking/payload/mod.rs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (dsluijk) [inline src/networking/payload/mod.rs]: Please remove this unnecessary empty line

File: src/networking/payload/mod.rs [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+// module trying to ensure feature parity with py-ipv8/ipv8/messaging/payload.py
+
```
Comment (dsluijk) [inline src/networking/payload/mod.rs]: Please remove this unnecessary empty line

File: src/networking/payload/mod.rs [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+
+pub mod payload;
+pub mod introductionrequestpayload;
+pub mod packet;
+pub mod bits;
+
+
```
Comment (dsluijk) [inline src/networking/payload/mod.rs]: Please remove this unnecessary empty line

File: src/networking/payload/packet.rs [hunk 11]
```diff
@@ -0,0 +1,430 @@
+
```
Comment (dsluijk) [inline src/networking/payload/packet.rs]: Please remove this unnecessary empty line

File: src/networking/payload/packet.rs [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+pub type PacketFormat = Vec<Vec<PacketFormatIdentifier>>;
+
+#[derive(Debug)]
+pub struct Packet {
+  data: Vec<u8>,
+}
+
+impl Packet {
+
```
Comment (dsluijk) [inline src/networking/payload/packet.rs]: Please remove this unnecessary empty line

File: src/networking/payload/packet.rs [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+    Packet {
+      data: data.to_vec(),
+    }
+  }
+
+  /// unpack a packet based on it's format into rust data types.
+  pub fn unpack(&self, format: PacketFormat) -> Option<Vec<Vec<PacketDataIdentifier>>> {
+    let mut finalres: Vec<Vec<PacketDataIdentifier>> = Vec::new();
+    let mut data = self.data.iter();
+
```
Comment (dsluijk) [inline src/networking/payload/packet.rs:92]: Please remove this unnecessary empty line

File: src/networking/payload/packet.rs [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+    if match data.next() {
+      Some(_) => true,
+      None => false,
+    } {
+      return None;
+    }
+
+    return Some(finalres);
+  }
+
```
Comment (dsluijk) [inline src/networking/payload/packet.rs]: Please remove this unnecessary empty line

File: src/networking/payload/payload.rs [hunk 15]
```diff
@@ -0,0 +1,17 @@
+use super::packet::Packet;
+
```
Comment (dsluijk) [inline src/networking/payload/payload.rs:2]: Please remove these unnecessary empty lines

File: src/networking/payload/payload.rs [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+
+
+
+
+pub trait Ipv8Payload {
+
```
Comment (dsluijk) [inline src/networking/payload/payload.rs]: Please remove this unnecessary empty line

File: src/networking/payload/payload.rs [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  // // fn serialize(&self) -> Vec<PackingIdentifier>;
+  // fn deserialize(data:Vec<PackingIdentifier>) -> Self;
+
```
Comment (dsluijk) [inline src/networking/payload/payload.rs]: Please remove this unnecessary empty line

File: src/networking/payload/payload.rs [hunk 18]
```diff
@@ -0,0 +1,17 @@
+use super::packet::Packet;
+
+
+
+
+
+pub trait Ipv8Payload {
+
+  // // fn serialize(&self) -> Vec<PackingIdentifier>;
```
Comment (dsluijk) [inline src/networking/payload/payload.rs]: Is this supposed to be commended out twice

File: src/networking/payload/payload.rs [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+  // fn deserialize(data:Vec<PackingIdentifier>) -> Self;
+
+
+  fn pack(&self) -> Packet;
+  fn unpack(packet:Packet) -> Self;
+
```
Comment (dsluijk) [inline src/networking/payload/payload.rs]: Please remove this unnecessary empty line

File: tests/addresstest.rs [hunk 20]
```diff
@@ -0,0 +1,8 @@
+use ipv8::networking::address;
+
```
Comment (dsluijk) [inline tests/addresstest.rs]: Please remove these unnecessary empty lines

**Review by dsluijk — CHANGES_REQUESTED** [review]

File: src/networking/payload/bits.rs (hunk 3 above)
Comment (jdonszelmann) [reply src/networking/payload/bits.rs]: makes the struct printable and comparable (all elements of the struct need to be equal)

File: src/networking/payload/bits.rs (hunk 4 above)
Comment (jdonszelmann) [reply src/networking/payload/bits.rs]: the bits struct is a construct used in py-ipv8 which i kept in mostly to keep their features similar. it's a struct containing booleans representing each bit in an u8. the from(u8) just creates such a struct from an u8

File: src/networking/payload/introductionrequestpayload.rs (hunk 5 above)
Comment (jdonszelmann) [reply src/networking/payload/introductionrequestpayload.rs:50]: i cant as the lower case versions are built in types. AFAIK an enum cant have types in it so i use the upper case version to signify which type there is. these vectors are basically format specifiers which tell the system which types to expect in a message

File: src/networking/payload/payload.rs (hunk 17 above)
Comment (jdonszelmann) [reply src/networking/payload/payload.rs]: whoops, forgot to remove those comments

File: src/networking/payload/payload.rs (hunk 18 above)
Comment (jdonszelmann) [reply src/networking/payload/payload.rs]: nope

File: src/networking/payload/bits.rs (hunk 4 above)
Comment (NULLx76) [reply src/networking/payload/bits.rs]: I would suggest adding a comment denoting this as it isn't self-evident

Comment (jdonszelmann) [reply src/networking/payload/bits.rs]: done that

File: src/networking/payload/introductionrequestpayload.rs (hunk 6 above)
Comment (jdonszelmann) [reply src/networking/payload/introductionrequestpayload.rs]: no not at all, i have only implemented unpacking, not the creation of payloads. These are some dummy values to make it compile.

**Review by dsluijk — APPROVED** [review]

**Pull request merged**